grails-app/services/org/pih/warehouse/data/LocationDataService.groovy

```diff
@@ -33,12 +34,17 @@ class LocationDataService {
         def locationNumbersToImport = command.data.collect {it.locationNumber}
 
         command.data.eachWithIndex { params, index ->
+            Location location = params.id ? Location.findById(params.id) : null
             Location parentLocation = params.parentLocation ? Location.findByName(params.parentLocation) : null
             LocationGroup locationGroup = params.locationGroup ? LocationGroup.findByName(params.locationGroup) : null
             LocationType locationType = params.locationType ? LocationType.findByName(params.locationType) : null
             List<Organization> organizations = params.organization ? Organization.findAllByCodeOrName(params.organization, params.organization) : null
 
-            if (Location.findByName(params.name)) {
+            if (params.id && !location) {
+                command.errors.reject("Row ${index + 1}: Id '${params.id}' do not exists in the system.")
+            }
+
+            if (Location.findByName(params.name) && !(params.id && params.name == location?.name)) {
```

Comment:
Holy smokes. Let's add a comment here to explain what's going on :)

Reply:
I would imagine that we also need to validate that we cannot create a location with the same location number as another location. There's a unique constraint on locationNumber but this doesn't work when parentLocation is null. So we probably need to create our own custom validation.

Also I don't know if I like the idea of using the id in the import / export. I think the location number should be our unique key. The one case I'm thinking of is someone uploading a spreadsheet with new locations, then immediately making an edit to that spreadsheet and uploading it again. Good UX would to allow the second import attempt because we have populated the location number and that's enough to qualify uniqueness.

Reply:
I've got validation for location number a bit lower.

Adding id field was specified in a ticket. Right now while there is no id and location number exists in a system than validation will not let to create new location.